File: leetcode1518/solution.cpp
class Solution {
public:
    int numWaterBottles(int numBottles, int numExchange) {
        /*
        Time: O( log n ) | Space: O( 1 )
        
        n: numBottles
        */
        int res = numBottles;

        while (numBottles >= numExchange) {
            res += numBottles / numExchange;
            numBottles = numBottles / numExchange + numBottles % numExchange;
        }

        return res;
    }
};

// class Solution {
// public:
//     int numWaterBottles(int numBottles, int numExchange) {
//         /*
//         Time: O( log n ) | Space: O( 1 )
        
//         n: numBottles
//         */
//         int res = 0;
//         int rem = numBottles % numExchange;

//         while (numBottles / numExchange > 0) {
//             res += numBottles - rem;
//             numBottles /= numExchange;
//             numBottles += rem;
//             rem = numBottles % numExchange;
//         }

//         res += rem;

//         return res;
//     }
// };

File: leetcode1518/solution.rs
impl Solution {
    pub fn num_water_bottles(mut num_bottles: i32, num_exchange: i32) -> i32 {
        /// Time: O(n) | Space: O(1)
        /// 
        /// where n is the number of bottles
        let mut res: i32 = num_bottles;

        while num_bottles >= num_exchange {
            res += num_bottles / num_exchange;
            num_bottles = num_bottles / num_exchange + num_bottles % num_exchange;
        }

        res
    }
}